// cdk-rds-sql/lib/database.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const constructs_1 = require("constructs");
const enum_1 = require("./enum");
const provider_1 = require("./provider");
class ImportedDatabase extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.databaseName = props.databaseName;
    }
}
class Database extends aws_cdk_lib_1.CustomResource {
    /**
     * Return a Database based upon name only. Use for importing existing databases.
     */
    static fromDatabaseName(scope, id, databaseName) {
        return new ImportedDatabase(scope, id, {
            databaseName: databaseName,
        });
    }
    constructor(scope, id, props) {
        // Check if using DSQL provider and forbid database creation
        if (props.provider.engine === provider_1.DatabaseEngine.DSQL) {
            throw new Error("Database creation is not supported with DSQL. DSQL always uses 'postgres' database.");
        }
        super(scope, id, {
            serviceToken: props.provider.serviceToken,
            properties: {
                Resource: enum_1.RdsSqlResource.DATABASE,
                ResourceId: props.databaseName,
                ...(props.provider.secret ? { SecretArn: props.provider.secret.secretArn } : {}),
                Owner: props.owner?.roleName,
            },
        });
        this.node.addDependency(props.provider);
        this.databaseName = props.databaseName;
        if (props.owner) {
            this.node.addDependency(props.owner);
        }
    }
}
exports.Database = Database;
_a = JSII_RTTI_SYMBOL_1;
Database[_a] = { fqn: "cdk-rds-sql.Database", version: "7.3.2" };
