// cdk-rds-sql/lib/iam-grant.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamGrant = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const enum_1 = require("./enum");
const provider_1 = require("./provider");
class IamGrant extends aws_cdk_lib_1.CustomResource {
    constructor(scope, id, props) {
        // IAM grants are only supported on DSQL
        if (props.provider.engine !== provider_1.DatabaseEngine.DSQL) {
            throw new Error("IAM grants are only supported with DSQL clusters. Use regular database authentication for RDS/Aurora clusters.");
        }
        super(scope, id, {
            serviceToken: props.provider.serviceToken,
            resourceType: "Custom::DsqlIamGrant",
            properties: {
                Resource: enum_1.RdsSqlResource.IAM_GRANT,
                RoleName: props.roleName,
                ResourceArn: props.resourceArn,
            },
        });
        this.node.addDependency(props.provider);
    }
}
exports.IamGrant = IamGrant;
_a = JSII_RTTI_SYMBOL_1;
IamGrant[_a] = { fqn: "cdk-rds-sql.IamGrant", version: "7.3.2" };
