// cdk-rds-sql/lib/provider.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Provider = exports.DatabaseEngine = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs_1 = require("fs");
const path = require("path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const dsql = require("aws-cdk-lib/aws-dsql");
const aws_ec2_1 = require("aws-cdk-lib/aws-ec2");
const iam = require("aws-cdk-lib/aws-iam");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const lambda = require("aws-cdk-lib/aws-lambda-nodejs");
const customResources = require("aws-cdk-lib/custom-resources");
const constructs_1 = require("constructs");
/**
 * Helper function to determine if a cluster is a DSQL cluster
 */
function isDsqlCluster(cluster) {
    return cluster instanceof dsql.CfnCluster;
}
/**
 * Supported database engines
 */
var DatabaseEngine;
(function (DatabaseEngine) {
    DatabaseEngine["POSTGRES"] = "postgres";
    DatabaseEngine["MYSQL"] = "mysql";
    DatabaseEngine["DSQL"] = "dsql";
})(DatabaseEngine || (exports.DatabaseEngine = DatabaseEngine = {}));
class Provider extends constructs_1.Construct {
    /**
     * Import an existing provider Lambda function
     */
    static fromProviderAttributes(scope, id, attrs) {
        return new ImportedProvider(scope, id, attrs);
    }
    constructor(scope, id, props) {
        super(scope, id);
        // Validate configuration
        const isDsql = isDsqlCluster(props.cluster);
        if (!isDsql && !props.secret) {
            throw new Error("Either secret (for traditional RDS) or cluster with DSQL must be provided");
        }
        if (!isDsql && !props.vpc) {
            throw new Error("VPC is required for traditional RDS databases");
        }
        if (isDsql && props.secret) {
            throw new Error("secret should not be provided when using DSQL cluster (uses IAM authentication)");
        }
        this.secret = props.secret;
        this.cluster = props.cluster;
        // Determine engine from cluster/instance instead of hardcoding
        if (isDsql) {
            // DSQL is always PostgreSQL-compatible
            this.engine = DatabaseEngine.DSQL;
        }
        else if ("clusterIdentifier" in props.cluster) {
            // It's a DatabaseCluster
            const clusterEngine = props.cluster.engine;
            this.engine =
                clusterEngine && clusterEngine.engineFamily === "MYSQL"
                    ? DatabaseEngine.MYSQL
                    : DatabaseEngine.POSTGRES;
        }
        else if ("instanceIdentifier" in props.cluster) {
            // It's a DatabaseInstance
            const instanceEngine = props.cluster.engine;
            this.engine =
                instanceEngine && instanceEngine.engineFamily === "MYSQL"
                    ? DatabaseEngine.MYSQL
                    : DatabaseEngine.POSTGRES;
        }
        else {
            // Fallback to postgres if engine hasn't been provided
            this.engine = DatabaseEngine.POSTGRES;
        }
        const functionName = "RdsSql" + slugify("28b9e791-af60-4a33-bca8-ffb6f30ef8c5");
        this.handler =
            aws_cdk_lib_1.Stack.of(this).node.tryFindChild(functionName) ??
                this.newCustomResourceHandler(scope, functionName, props);
        const provider = new customResources.Provider(this, "RdsSql", {
            onEventHandler: this.handler,
        });
        this.serviceToken = provider.serviceToken;
        // Handle database connection setup
        if (isDsql) {
            // For DSQL, grant IAM permissions instead of VPC security groups
            const dsqlCluster = props.cluster;
            this.handler.addToRolePolicy(new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: ["dsql:DbConnectAdmin"],
                resources: [dsqlCluster.attrResourceArn],
            }));
        }
        else {
            // Traditional RDS setup with security groups and secrets
            this.secret.grantRead(this.handler);
            if (this.secret.encryptionKey) {
                // It seems we need to grant explicit permission
                this.secret.encryptionKey.grantDecrypt(this.handler);
            }
            if (props.cluster.connections.securityGroups.length === 0) {
                throw new Error("Cluster does not have a security group.");
            }
            else {
                const securityGroup = props.cluster.connections.securityGroups[0];
                this.handler.node.defaultChild?.node.addDependency(securityGroup);
            }
        }
        this.node.addDependency(props.cluster);
    }
    newCustomResourceHandler(scope, id, props) {
        const isDsql = isDsqlCluster(props.cluster);
        const handlerDir = path.join(__dirname, "handler");
        const index_ts = path.join(handlerDir, "index.ts");
        const index_js = path.join(handlerDir, "index.js");
        let entry;
        if ((0, fs_1.existsSync)(index_ts)) {
            entry = index_ts;
        }
        else if ((0, fs_1.existsSync)(index_js)) {
            entry = index_js;
        }
        else {
            // Ugly hack to support SST (possibly caused by my hack to make SST work with CommonJS libraries)
            entry = path.join(path.dirname(process.env.npm_package_json || process.cwd()), "node_modules/cdk-rds-sql/lib/handler/index.js");
        }
        let ssl_options;
        if (props.ssl !== undefined && !props.ssl) {
            ssl_options = {
                SSL: JSON.stringify(props.ssl),
            };
        }
        const logger = props.logger ?? false;
        // Build environment variables
        const environment = {
            LOGGER: logger.toString(),
            ...ssl_options,
        };
        // Add DSQL-specific environment variables
        if (isDsql) {
            const dsqlCluster = props.cluster;
            const clusterId = dsqlCluster.attrIdentifier;
            const region = aws_cdk_lib_1.Stack.of(scope).region;
            environment.DSQL_ENDPOINT = `${clusterId}.dsql.${region}.on.aws`;
            environment.DSQL_PORT = "5432";
        }
        else if (props.secret) {
            // Add secret ARN to environment for traditional RDS
            environment.SECRET_ARN = props.secret.secretArn;
        }
        const deleteParameterPolicy = new iam.PolicyStatement({
            actions: ["ssm:DeleteParameter"],
            resources: [
                `arn:aws:ssm:${aws_cdk_lib_1.Stack.of(scope).region}:${aws_cdk_lib_1.Stack.of(scope).account}:parameter/*`,
            ],
            conditions: {
                StringEquals: {
                    "ssm:ResourceTag/created-by": "cdk-rds-sql",
                },
            },
        });
        const fn = new lambda.NodejsFunction(scope, id, {
            ...props.functionProps,
            // Only configure VPC for traditional RDS
            ...(isDsql
                ? {}
                : {
                    vpc: props.vpc,
                    vpcSubnets: props.vpcSubnets ?? {
                        subnetType: aws_ec2_1.SubnetType.PRIVATE_ISOLATED,
                    },
                }),
            entry: entry,
            runtime: aws_lambda_1.Runtime.NODEJS_22_X,
            timeout: props.timeout ?? props.functionProps?.timeout ?? aws_cdk_lib_1.Duration.seconds(300),
            bundling: {
                // Include the migrations directory in the bundle
                commandHooks: {
                    beforeBundling(_, outputDir) {
                        return [
                            `curl --silent -fL https://truststore.pki.rds.amazonaws.com/global/global-bundle.pem -o ${path.join(outputDir, "global-bundle.pem")}`,
                        ];
                    },
                    afterBundling() {
                        return [];
                    },
                    beforeInstall() {
                        return [];
                    },
                },
            },
            environment,
            initialPolicy: [
                deleteParameterPolicy,
                ...(props.functionProps?.initialPolicy ?? []),
            ],
        });
        // Only configure security groups for traditional RDS (not DSQL)
        if (!isDsql &&
            (!props.functionProps?.securityGroups ||
                props.functionProps?.securityGroups.length === 0)) {
            const rdsCluster = props.cluster;
            rdsCluster.connections.allowDefaultPortFrom(fn, "Allow the rds sql handler to connect to db");
        }
        return fn;
    }
}
exports.Provider = Provider;
_a = JSII_RTTI_SYMBOL_1;
Provider[_a] = { fqn: "cdk-rds-sql.Provider", version: "7.3.2" };
class ImportedProvider extends constructs_1.Construct {
    constructor(scope, id, attrs) {
        super(scope, id);
        // Validate that either functionArn or functionName is provided
        if (!attrs.functionArn && !attrs.functionName) {
            throw new Error("Either functionArn or functionName must be provided");
        }
        if (attrs.functionArn && attrs.functionName) {
            throw new Error("Provide either functionArn or functionName, not both");
        }
        // Import the existing Lambda function
        this.handler = attrs.functionArn
            ? aws_lambda_1.Function.fromFunctionArn(this, "Handler", attrs.functionArn)
            : aws_lambda_1.Function.fromFunctionName(this, "Handler", attrs.functionName);
        // Derive serviceToken by wrapping in custom resource provider
        const provider = new customResources.Provider(this, "Provider", {
            onEventHandler: this.handler,
        });
        this.serviceToken = provider.serviceToken;
        this.engine = attrs.engine;
        this.secret = undefined; // Imported providers get secret from environment
        this.cluster = attrs.cluster; // Optional cluster for role creation
    }
}
function slugify(x) {
    return x.replace(/[^a-zA-Z0-9]/g, "");
}
