// cdk-rds-sql/lib/role.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Role = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const iam = require("aws-cdk-lib/aws-iam");
const aws_secretsmanager_1 = require("aws-cdk-lib/aws-secretsmanager");
const ssm = require("aws-cdk-lib/aws-ssm");
const constructs_1 = require("constructs");
const enum_1 = require("./enum");
const provider_1 = require("./provider");
const role_custom_resource_1 = require("./role.custom-resource");
// Private Parameters construct (not exported)
class Parameters extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        // Create parameters for each key-value pair
        Object.entries(props.paramData).forEach(([key, value]) => {
            if (value !== undefined) {
                new ssm.StringParameter(this, `Parameter-${key}`, {
                    parameterName: `${props.parameterPrefix}${key}`,
                    stringValue: value.toString(),
                });
            }
        });
        // For password, use the existing provider to store it in SSM
        const passwordParameterName = `${props.parameterPrefix}password`;
        const password_parameter = new aws_cdk_lib_1.CustomResource(this, "PasswordParameter", {
            serviceToken: props.providerServiceToken,
            properties: {
                SecretArn: props.secretArn,
                Resource: enum_1.RdsSqlResource.PARAMETER_PASSWORD,
                PasswordArn: props.passwordArn,
                ParameterName: passwordParameterName,
            },
        });
        password_parameter.node.addDependency(props.provider);
        const paramArn = `arn:aws:ssm:${aws_cdk_lib_1.Stack.of(this).region}:${aws_cdk_lib_1.Stack.of(this).account}:parameter${passwordParameterName.startsWith("/") ? "" : "/"}${passwordParameterName}`;
        props.provider.handler.addToRolePolicy(new iam.PolicyStatement({
            actions: ["ssm:PutParameter", "ssm:AddTagsToResource", "ssm:GetParameters"],
            resources: [paramArn],
        }));
    }
}
class Role extends constructs_1.Construct {
    constructor(scope, id, props) {
        if (props.provider.engine !== provider_1.DatabaseEngine.DSQL) {
            if (props.database && props.databaseName) {
                throw "Specify either database or databaseName";
            }
            if (!props.database && !props.databaseName) {
                // If neither is specified, we might need a default or throw an error depending on desired behavior.
                // For now, let's assume it's allowed but the secret won't have a dbname.
                // If it should be required, uncomment the line below:
                throw "Specify either database or databaseName";
            }
        }
        super(scope, id);
        // DSQL doesn't use secrets - it always uses IAM authentication
        if (props.provider.engine !== provider_1.DatabaseEngine.DSQL) {
            // For imported providers without cluster details, provide helpful error message
            if (!props.provider.cluster) {
                throw new Error("Role creation requires cluster information. When importing a provider with " +
                    "Provider.fromProviderAttributes(), include the 'cluster' property if you plan " +
                    "to create new roles. Alternatively, use existing roles created with the original provider.");
            }
            // For RDS/Aurora clusters and instances, get endpoint details
            const host = props.provider.cluster.clusterEndpoint
                ? props.provider.cluster.clusterEndpoint.hostname
                : props.provider.cluster.instanceEndpoint.hostname;
            const port = props.provider.cluster.clusterEndpoint
                ? props.provider.cluster.clusterEndpoint.port
                : props.provider.cluster.instanceEndpoint.port;
            const identifier = props.provider.cluster.clusterIdentifier
                ? props.provider.cluster.clusterIdentifier
                : props.provider.cluster.instanceIdentifier;
            const secretTemplate = {
                dbClusterIdentifier: identifier,
                engine: props.provider.engine,
                host: host,
                port: port,
                username: props.roleName,
                dbname: props.database ? props.database.databaseName : props.databaseName,
            };
            this.secret = new aws_secretsmanager_1.Secret(this, "Secret", {
                secretName: props.secretName,
                encryptionKey: props.encryptionKey,
                description: `Generated secret for ${props.provider.engine} role ${props.roleName}`,
                ...(props.enableIamAuth
                    ? {
                        // For IAM auth, create secret without password generation
                        secretStringTemplate: JSON.stringify(secretTemplate),
                    }
                    : {
                        // For password auth, generate password
                        generateSecretString: {
                            passwordLength: 30, // Oracle password cannot have more than 30 characters
                            secretStringTemplate: JSON.stringify(secretTemplate),
                            generateStringKey: "password",
                            excludeCharacters: " %+~`#$&*()|[]{}:;<>?!'/@\"\\",
                        },
                    }),
                removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
            });
            // Create Parameters if parameterPrefix is provided
            if (props.parameterPrefix) {
                const paramData = {
                    dbClusterIdentifier: identifier,
                    engine: props.provider.engine,
                    host: host,
                    port: port,
                    username: props.roleName,
                    dbname: props.database ? props.database.databaseName : props.databaseName,
                };
                new Parameters(this, "Parameters", {
                    secretArn: props.provider.secret?.secretArn || "",
                    parameterPrefix: props.parameterPrefix,
                    passwordArn: props.enableIamAuth ? "" : this.secret.secretArn,
                    providerServiceToken: props.provider.serviceToken,
                    provider: props.provider,
                    paramData,
                });
            }
        }
        const role = new role_custom_resource_1.Role(this, "PostgresRole", {
            provider: props.provider,
            roleName: props.roleName,
            passwordArn: props.enableIamAuth || props.provider.engine === provider_1.DatabaseEngine.DSQL
                ? ""
                : this.secret.secretArn,
            database: props.database,
            databaseName: props.databaseName,
            enableIamAuth: props.enableIamAuth || props.provider.engine === provider_1.DatabaseEngine.DSQL,
        });
        if (this.secret) {
            role.node.addDependency(this.secret);
            this.secret.grantRead(props.provider.handler);
            if (this.secret.encryptionKey) {
                // It seems we need to grant explicit permission
                this.secret.encryptionKey.grantDecrypt(props.provider.handler);
            }
        }
        this.roleName = props.roleName;
    }
}
exports.Role = Role;
_a = JSII_RTTI_SYMBOL_1;
Role[_a] = { fqn: "cdk-rds-sql.Role", version: "7.3.2" };
