// cdk-rds-sql/lib/role.custom-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Role = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const enum_1 = require("./enum");
class Role extends aws_cdk_lib_1.CustomResource {
    constructor(scope, id, props) {
        if (props.database && props.databaseName) {
            throw "Specify either database or databaseName";
        }
        super(scope, id, {
            serviceToken: props.provider.serviceToken,
            properties: {
                Resource: enum_1.RdsSqlResource.ROLE,
                ResourceId: props.roleName,
                ...(props.provider.secret ? { SecretArn: props.provider.secret.secretArn } : {}),
                PasswordArn: props.passwordArn,
                DatabaseName: props.database ? props.database.databaseName : props.databaseName,
                EnableIamAuth: props.enableIamAuth,
            },
        });
        this.node.addDependency(props.provider);
        if (props.database)
            this.node.addDependency(props.database);
    }
}
exports.Role = Role;
