// cdk-rds-sql/lib/schema.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schema = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const enum_1 = require("./enum");
class Schema extends aws_cdk_lib_1.CustomResource {
    constructor(scope, id, props) {
        super(scope, id, {
            serviceToken: props.provider.serviceToken,
            properties: {
                Resource: enum_1.RdsSqlResource.SCHEMA,
                ResourceId: props.schemaName,
                ...(props.provider.secret ? { SecretArn: props.provider.secret.secretArn } : {}),
                DatabaseName: props.database ? props.database.databaseName : undefined,
                RoleName: props.role ? props.role.roleName : undefined,
            },
        });
        this.node.addDependency(props.provider);
        this.schemaName = props.schemaName;
        if (props.database)
            this.node.addDependency(props.database);
        if (props.role)
            this.node.addDependency(props.role);
    }
}
exports.Schema = Schema;
_a = JSII_RTTI_SYMBOL_1;
Schema[_a] = { fqn: "cdk-rds-sql.Schema", version: "7.3.2" };
