// cdk-rds-sql/lib/sql.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sql = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const enum_1 = require("./enum");
class Sql extends aws_cdk_lib_1.CustomResource {
    constructor(scope, id, props) {
        super(scope, id, {
            serviceToken: props.provider.serviceToken,
            properties: {
                Resource: enum_1.RdsSqlResource.SQL,
                ...(props.provider.secret ? { SecretArn: props.provider.secret.secretArn } : {}),
                DatabaseName: props.database ? props.database.databaseName : undefined,
                Statement: props.statement,
                Rollback: props.rollback,
            },
        });
        this.node.addDependency(props.provider);
        if (props.database)
            this.node.addDependency(props.database);
    }
}
exports.Sql = Sql;
_a = JSII_RTTI_SYMBOL_1;
Sql[_a] = { fqn: "cdk-rds-sql.Sql", version: "7.3.2" };
